teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,483 @@
+ }
+```
+
+For the `outputs`, we are looking at the following support, as these are the artifact types supported in Artifact Registries. However, the list should definitely be expanded as this doc is spreading the discussions in the communities to see which should be prioritized. Currently, we have the following under considerations: Python, Maven, Go, NodeJS and OCI images. Note that OCI images are the ones we currently support.
+
+#### Concrete Use Cases
+
+All below examples will be generating in-toto attestation.
+ 1. Use Git commits as sources, and generates in-toto provenance for a TaskRun that builds an image.
+   
+    A Task example **without** structured results and params support can be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-GIT_COMMIT
+          type: string
+          description: git commit sha
+        - name: CHAINS-GIT_URL
+          type: string
+          description: git commit url
+        - name: FOO_IMAGE
+          type: string
+          description: artifact image
+      ...
+      results:
+        - name: IMAGE_DIGEST
+          description: digest of image
+        - name: IMAGE_URL
+          description: url of image
+    ```
+    A Task example `with` structured results support will be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      ...
+      results:
+        - name: ARTIFACT-INPUTS_git-vcs
+          type: object
+          description: |
+            The source distribution
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
+          properties:
+            uri:
+              type: string
+            digest:
+              type: string
+        - name: ARTIFACT-OUTPUTS_oci_image
+          type: object
+          description: N/A
+          properties:
+            uri:
+              type: string
+            digest:
+              type: string
+    ```
+    And its corresponding TaskRun example will be like:
+    ``` yaml
+    TaskRun:
+      ...
+      results:
+        - name: ARTIFACT-INPUTS_git-vcs
+          value:
+            uri: git+https://github.com/foo/bar.git
+            digest: sha256:abc
+        - name: ARTIFACT-OUTPUTS_oci-image
+          value:
+            uri: gcr.io/somerepo/someimage
+            digest: sha512:abc
+    ```
+The generated intoto provenance will contain the following:
+```
+subjects: [{"name": "gcr.io/somerepo/someimage", "digest": {"sha256": "abc"}}]
+...
+materials: [{
+  "uri": "git+https://github.com/foo/bar.git",
+  "digest": {"sha1": "abc..."}
+}]
+```
+
+2. Use Perforce and images as sources, and generates in-toto provenance for a TaskRun that builds a Maven package.
+``` yaml
+results:
+  - name: ARTIFACT-INPUTS_perforce-vcs
+    value:
+      uri: http://myp4web:8080/depot/main/atlas/
+      digest: sha256:abc
+  - name: ARTIFACT-OUTPUTS_maven-pkg
```

Comment:
Do the artifacts have to provide the "type" (i.e. maven, perforce etc.) as part of the name? Or is that just for illustrating these examples?

Reply:
Thanks for bringing this up! Yeah probably this is not really clear from the examples. This is explained more in the proposal part that type doesn't need to be there in the naming.  This is just a demonstration of how one of these types are used